kv: Use strict types for common fields

andrewbaptist opened the pull request:
This PR introduces 3 new typed fields in mvcc.go:
RaftTerm, RaftIndex and LeaseSequence. These fields were previously just unit64 throughout the code and this made the code harder to read and risked incorrect conversions.

Epic: none

Release note: None

Commit:
kv: Use strict types for common fields

This PR introduces 3 new typed fields to the kvpb package: RaftTerm,
RaftIndex and LeaseAppliedIndex. These fields were previously just
unit64 throughout the code and this made the code harder to read and
risked incorrect conversions.

It also moves internal_raft.proto into the kvserver package as it is
no longer accessed outside of kv.

Epic: none

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/100181)
<!-- Reviewable:end -->

andrewbaptist commented:
This was a 95% mechanical change and seemed pretty useful.

There are three things that I don't love about this PR but couldn't easily change:

1) The etc.d raft library doesn't strongly type the Term or Index, so it's necessary to cast for all the places that interact with that. There aren't too many of them, and they are usually very clear what the type is, so I don't think this is a big problem.
2) I needed to put the type definitions in the `enginepb` library so they could be used everywhere. I originally planned on putting them in `roachpb` but that doesn't work because of `RangeAppliedState` which is in `mvcc3.proto` 
3) The `LeaseSequence` was inconsistently a uint64 or a `int64`. I made the type `int64` to make sure there weren't any problems with negative numbers, but it probably doesn't matter which one it is. 

Let me know if you have any thoughts on this. 

Finally I put a few reviewers on here, but I don't expect everyone to look at every line. I'm not sure the best way to tag it.

andrewbaptist commented:
Also if we decide to make this change should be backported? It feels pretty low risk and could avoid some merge conflicts on backports. However, it probably isn't that important to do as the conflicts would be trivial to resolve. There are no serialization format changes to any protobufs.

andrewbaptist commented:
> Let's move RangeAppliedState and LeaseSequence into kvserverpb/state.{proto,go}, and the Raft types into kvserverpb/raft.{proto,go}. None of this has anything to do with MVCC. Hopefully that'll be straightforward.

I was able to move them into kvserverpb at least, but it would have taken a fair bit more rearranging to move then into the more specific packages. This would probably be a good thing to do as a follow-on PR later.

> Since we're doing all of this work anyway, let's just flip it to uint64. I think a negative lease sequence would have stuff blowing up anyway, and they should be wire-compatible (but I've been wrong about Protobuf changes before).

I couldn't do this without breaking `TestBelowRaftProtosDontChange`. I tried a couple of variations of this, but it wasn't possible to make it a uint64 but serialize as a int64 either since it hit a negative check. 

> I'd prefer not to. There's only downside to users, if we get something wrong. 

Sounds good - I won't plan it for backport. The only risk I saw was making some PRs harder to backport, but I think that is not overly important.

erikgrinaker reviewed (commented):
> I was able to move them into kvserverpb at least, but it would have taken a fair bit more rearranging to move then into the more specific packages. This would probably be a good thing to do as a follow-on PR later.

That's fine, we can clean it up later. I take it we were forced to move the index/term types into `roachpb` for similar dependency reasons?

>I couldn't do this without breaking `TestBelowRaftProtosDontChange`. I tried a couple of variations of this, but it wasn't possible to make it a uint64 but serialize as a int64 either since it hit a negative check.

I think that's just because the test uses Protobuf-generated `Populate*` functions that randomly populate messages using a fixed seed. When we change a field from int64 to uint64 it can both change the randomly chosen values, and possibly also inject additional calls to the RNG such as `if r.Intn(2) == 0 { this.Sequence *= -1 }`. It can therefore change both the randomly generated values and the sequence, but that doesn't mean that the messages aren't wire-compatible.

However, when I tried changing it to uint64 and running `TestBelowRaftProtosDontChange` it passed. 🤷‍♀️

> Reviewed 40 of 119 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/100181#-:-NSBXHwbBVvIKN6FKRjg:bo7mxq2)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andrewbaptist, @nvanbenschoten, @pavelkalinnikov, @RaduBerinde, and @tbg)

___
*[`-- commits` line 4 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/100181#-NSBOUgsDsAYgsrznql8):*
Needs updating.
___
*[`pkg/kv/kvpb/api.proto` line 2102 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/100181#-NSBOpd6ALOwL5upGvjY:-NSBOpd6ALOwL5upGvjZ:bb6ngp1) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e3a1118ab11583be0a9b5d224c1c1fdd5e1d93a/pkg/kv/kvpb/api.proto#L2102)):*
> ```protobuf
>   // applied index of the CPut that left an intent on the local copy of the
>   // right-hand range descriptor.
>   int64 lease_applied_index = 4  [(gogoproto.casttype) = "github.com/cockroachdb/cockroach/pkg/roachpb.LeaseSequence"];
> ```

The lease applied index is not a lease sequence number, so this type isn't appropriate. Didn't we already have a separate type for it?

The lease sequence number distinguishes different leases from each other. It is incremented when we acquire a new lease.

The lease applied index is used to order individual Raft proposals, for replay protection. It is incremented for every Raft proposal:

https://github.com/cockroachdb/cockroach/blob/6e3a1118ab11583be0a9b5d224c1c1fdd5e1d93a/pkg/kv/kvserver/replica_proposal_buf.go#L767
___
*[`pkg/kv/kvserver/kvserverpb/state.proto` line 40 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/100181#-NSBWSes9deFbDQtHXHa:-NSBWSes9deFbDQtHXHb:b-pn0lty) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e3a1118ab11583be0a9b5d224c1c1fdd5e1d93a/pkg/kv/kvserver/kvserverpb/state.proto#L40)):*
> ```protobuf
>   uint64 raft_applied_index = 1 [(gogoproto.casttype) = "github.com/cockroachdb/cockroach/pkg/roachpb.RaftIndex"];
>   // The highest (and last) lease index applied to the state machine.
>   int64 lease_applied_index = 2  [(gogoproto.casttype) = "github.com/cockroachdb/cockroach/pkg/roachpb.LeaseSequence"];
> ```

As mentioned elsewhere, this and other instances should be a LeaseAppliedIndex not a LeaseSequence.


<!-- Sent from Reviewable.io -->

andrewbaptist commented:
Thanks! I didn't realize the difference between LeaseSequence and LeaseAppliedIndex until now. This makes more sense. 

I updated to use uint64 everywhere and it is passing the tests now. I must have been doing something strange before, but it seems like you were right that we can do this. 

Finally, I wanted to move all these fields into kv so I looked at what was necessary to do that. It turned out that only the `internal_raft.proto` was holding this up, and that didn't make sense in `roachpb` anyway. I moved that over and updated the PR to have the fields in better locations.

erikgrinaker commented:
Thanks! I won't have time for another pass before OOO and breather week, but I suppose we'd want to let this sit until the 23.1 backport bonanza calms down anyway.

andrewbaptist commented:
Yep, we can let this sit a little. There isn't a rush to get this in. It shouldn't be too hard to keep this up-to-date with master. At least it is now all green.

andrewbaptist commented:
@erikgrinaker if you get a chance next week can you review this again? Would be nice to merge to master now that backports to 23.1 have slowed down.

tbg commented on pkg/kv/kvpb/data.go:
```diff
@@ -151,3 +151,25 @@ func TransactionRefreshTimestamp(pErr *Error) (bool, hlc.Timestamp) {
+type RaftIndex uint64
+
+// SafeValue implements the redact.SafeValue interface.
+func (s RaftTerm) SafeValue()  {}
```
nit: it's unusual to implement this "not next to" the receiver.

tbg commented on pkg/kv/kvserver/closedts/ctpb/service.go:
```diff
@@ -16,23 +16,20 @@ import (
 )
 
+// Used within the RangeUpdate method.
+var _ kvpb.LeaseAppliedIndex
```
If you need `kvpb` imported, use a `_ "github.com/cockroachdb/cockroach/pkg/kv/kvpb"` import instead.

tbg commented on pkg/kv/kvserver/loqrecovery/loqrecoverypb/recovery.go:
```diff
@@ -25,6 +26,9 @@ import (
 type RecoveryKey roachpb.RKey
 
+// Needed for recovery.proto.
+var _ kvpb.RaftIndex
```
Ditto

tbg reviewed (approved):
LGTM

Don't bother addressing the import nits in this PR if it's green, can be done as a follow-up PR.

andrewbaptist reviewed (commented):
No problem - all easy to address. TFTY

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/100181#-:-NU1tHse4T6I8GBlHexf:bmhew15)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker, @nvanbenschoten, @pavelkalinnikov, @RaduBerinde, and @tbg)

___
*[`-- commits` line 4 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/100181#-NSBOUgsDsAYgsrznql8):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

Needs updating.
</blockquote></details>

Updated - thanks.
___
*[`pkg/kv/kvpb/api.proto` line 2102 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/100181#-NSBOpd6ALOwL5upGvjY:-NU1tUeGA-fIEA8OKXUi:b-bq782o) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e3a1118ab11583be0a9b5d224c1c1fdd5e1d93a/pkg/kv/kvpb/api.proto#L2102)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

The lease applied index is not a lease sequence number, so this type isn't appropriate. Didn't we already have a separate type for it?

The lease sequence number distinguishes different leases from each other. It is incremented when we acquire a new lease.

The lease applied index is used to order individual Raft proposals, for replay protection. It is incremented for every Raft proposal:

https://github.com/cockroachdb/cockroach/blob/6e3a1118ab11583be0a9b5d224c1c1fdd5e1d93a/pkg/kv/kvserver/replica_proposal_buf.go#L767
</blockquote></details>

Fixed. I changed to separate these back out.
___
*[`pkg/kv/kvpb/data.go` line 174 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/100181#gh-1177869966:-NU1tkPh84Atv48QNPI9:b-icwirf) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5a900520b4fe5d7524c1a456b4c5085299e44cf/pkg/kv/kvpb/data.go#L174)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

nit: it's unusual to implement this "not next to" the receiver.
</blockquote></details>

Fixed, moved next to the definitions.
___
*[`pkg/kv/kvserver/closedts/ctpb/service.go` line 25 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/100181#gh-1177870894:-NU1u24HBqJ1Rb76GZ-D:biozj62) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5a900520b4fe5d7524c1a456b4c5085299e44cf/pkg/kv/kvserver/closedts/ctpb/service.go#L25)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

If you need `kvpb` imported, use a `_ "github.com/cockroachdb/cockroach/pkg/kv/kvpb"` import instead.
</blockquote></details>

Done. Thanks for the tip!
___
*[`pkg/kv/kvserver/kvserverpb/state.proto` line 40 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/100181#-NSBWSes9deFbDQtHXHa:-NU1u58DBuEb5GukG1rg:b499q7j) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e3a1118ab11583be0a9b5d224c1c1fdd5e1d93a/pkg/kv/kvserver/kvserverpb/state.proto#L40)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

As mentioned elsewhere, this and other instances should be a LeaseAppliedIndex not a LeaseSequence.
</blockquote></details>

Changed
___
*[`pkg/kv/kvserver/kvserverpb/state.proto` line 218 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/100181#gh-1177872169:-NU1uGt33jcBVoCeqve2:buxrq24) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5a900520b4fe5d7524c1a456b4c5085299e44cf/pkg/kv/kvserver/kvserverpb/state.proto#L218)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Moved from elsewhere with just casttype adjustments, right?
</blockquote></details>

Correct. This was in roachpb before, but more appropriate here.
___
*[`pkg/kv/kvserver/loqrecovery/loqrecoverypb/recovery.go` line 30 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/100181#gh-1177872505:-NU1v5j184V_3qoTwaBp:b-886gsx) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5a900520b4fe5d7524c1a456b4c5085299e44cf/pkg/kv/kvserver/loqrecovery/loqrecoverypb/recovery.go#L30)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Ditto
</blockquote></details>

Fixed


<!-- Sent from Reviewable.io -->

erikgrinaker reviewed (approved):
Thanks for the cleanup! Reviewed all of the Protobuf changes, LGTM.

For next time, it'd be nice to split this out into a few commits (e.g. file reorganization and Protobuf message changes), to more easily identify the changes that need particular scrutiny. But don't spend time on that now.

andrewbaptist commented:
bors r=erikgrinaker 

TFTR - I'll split it out any future cleanups.
